src/Elliptic/DiscontinuousGalerkin/NumericalFluxes/FirstOrderInternalPenalty.hpp

```diff
@@ -0,0 +1,379 @@
+ * penalty flux (see e.g. \cite HesthavenWarburton, section 7.2, or
+ * \cite Arnold2002)
+ *
+ * \f{align}
+ * n_i F^{*i}_u = n_i v_i^* = \frac{1}{2} n_i \left(\partial_i u^\mathrm{int} +
+ * \partial_i u^\mathrm{ext}\right) - \sigma \left(u^\mathrm{int} -
+ * u^\mathrm{ext}\right)
+ * n_i F^{*i}_{v_j} = n_j u^* = \frac{1}{2} n_j \left(u^\mathrm{int} +
+ * u^\mathrm{ext}\right)
+ * \f}
+ *
+ * where a sum over repeated indices is assumed, since the equation is
+ * formulated on a Euclidean geometry.
+ *
+ * This generalization of the internal penalty flux is based on unpublished work
+ * by Nils L. Fischer (nils.fischer@aei.mpg.de).
+ *
+ * The penalty factor \f$\sigma\f$ is responsible for removing zero eigenmodes
+ * and impacts the conditioning of the linear operator to be solved. It can be
+ * chosen as \f$\sigma=C\frac{N_\mathrm{points}^2}{h}\f$ where
+ * \f$N_\mathrm{points}\f$ is the number of collocation points (i.e. the
+ * polynomial degree plus 1), \f$h\f$ is a measure of the element size in
+ * inertial coordinates and \f$C\leq 1\f$ is a free parameter (see e.g.
+ * \cite HesthavenWarburton, section 7.2).
+ */
+template <size_t Dim, typename FluxesComputerTag, typename FieldTagsList,
+          typename AuxiliaryFieldTagsList,
+          typename FluxesComputer = db::item_type<FluxesComputerTag>,
+          typename FluxesArgs = typename FluxesComputer::argument_tags>
+struct FirstOrderInternalPenalty;
+
+template <size_t Dim, typename FluxesComputerTag, typename... FieldTags,
+          typename... AuxiliaryFieldTags, typename FluxesComputer,
+          typename... FluxesArgs>
+struct FirstOrderInternalPenalty<Dim, FluxesComputerTag,
+                                 tmpl::list<FieldTags...>,
+                                 tmpl::list<AuxiliaryFieldTags...>,
+                                 FluxesComputer, tmpl::list<FluxesArgs...>> {
+ private:
+  using fluxes_computer_tag = FluxesComputerTag;
+
+  template <typename Tag>
+  struct NormalDotDivFlux : db::PrefixTag, db::SimpleTag {
+    using tag = Tag;
+    using type = TensorMetafunctions::remove_first_index<typename Tag::type>;
+  };
+
+  template <typename Tag>
+  struct NormalDotNormalDotFlux : db::PrefixTag, db::SimpleTag {
+    using tag = Tag;
+    using type = TensorMetafunctions::remove_first_index<typename Tag::type>;
+  };
+
+ public:
+  struct PenaltyParameter {
+    using type = double;
+    // Currently this is used as the full prefactor to the penalty term. When it
+    // becomes possible to communicate non-tensors (the element size $h$ and
+    // the number of collocation points $p$ on either side of the mortar), this
+    // should be changed to be just the parameter multiplying
+    // $\frac{N_\mathrm{points}^2}{h}$. Currently, the `packaged_data` is a
+    // `Variables` which can only hold tensors.
+    static constexpr OptionString help = {
+        "The prefactor to the penalty term of the flux."};
+  };
+  using options = tmpl::list<PenaltyParameter>;
```

Comment:
If PenaltyParameter is an option, then doesn't it have to be the same for all elements? Does this mean that this assumes identical elements on both sides of the mortar? If so, the help text should say so

Reply:
Right now it is indeed just a number from the input file that's being used for all element faces. It doesn't mean the elements must be identical. The number just needs to be large enough so that the scheme is stable everywhere. For non-uniform meshes this means it is an overestimate where elements are large or polynomial degrees are small, but this only affects the conditioning of the scheme, i.e. it will converge slower but to the same solution. Once I can communicate the element size and the polynomial degree across elements I can compute the penalty dynamically, which will improve the conditioning on non-uniform meshes.

I added some of these details to the cpp comment.